Switch to activation-order for Layers

algernon opened the pull request:
Previously, we used index-ordering for layers, meaning, we looked keys up based on the index of active layers. This turned out to be confusing, and in many cases, limiting, since we couldn't easily shift to a lower layer from a higher one. As such, index-ordering required careful planning of one's layers, and a deeper understanding of the system.

This patch switches us to activation-ordering: the layer subsystem now keeps track of the order in which layers are activated, and uses that order to look keys up, instead of the index of layers. This makes it easier to understand how the system works, and allows us to shift to lower layers too.

It does require a bit more resources, since we can't just store a bitmap of active layers, but need 32 bytes to store the order. We still keep the bitmap, to make `Layer.isActive()` fast: looking up a bit in the bitmap is more efficient than walking the active layer array, and this function is often used in cases where speed matters.

As a side effect of the switch, a number of methods were deprecated, and similar ones with more appropriate names were introduced. See the updated `UPGRADING.md` document for more details.

Plugins that used the deprecated methods were updated to use the new ones.

Fixes #857.

gedankenexperimenter commented:
I mentioned this already in #857, but it bears repeating: this change will introduce a new race condition.

If I define two layers (`WHOLE` & `PARTIAL`) which are activated by holding layer shift keys, only one of which contains transparent keys, I will get non-deterministic behaviour when activating both layers simultaneously. If the `PARTIAL` layer gets triggered first, none of its keys will be visible while the two layer keys are held. If the `WHOLE` layer shift wins the race, however, `PARTIAL`'s non-transparent keys will be functional.

I don't think it's unreasonable to accept this race condition as necessary evil, but I do want to make sure that it gets considered before you guys make the change.

algernon commented:
Since GitHub was having issues before, I forgot to mention: this change results in +32 RAM and ~+106 PROGMEM usage (based on Model01-Firmware).

> I mentioned this already in #857, but it bears repeating: this change will introduce a new race condition.

Yes, that has been considered, and I believe it is an acceptable compromise. I'd say it's even the expected behaviour: the most recently activated layer will shadow layers further down the stack on all non-transparent positions. Personally, I find this much less confusing than the status quo, where if `PARTIAL` has a higher index than `WHOLE`, it would still override `WHOLE`, no matter the order of activation.

The "race" is only a race if both are pressed at the same time (as in, they both get registered in the same scan cycle). That - I think - would be a pretty rare case.

But it is worth documenting nevertheless.

It is also worth documenting (thanks @obra!) that you can't activate the same layer twice with this setup. We can't have a stack of `0, 1, 2, 1`.

algernon commented:
As this is a rather invasive change, it needs careful testing and review. Furthermore, because the Keyboardio Atreus mass production is about to happen, lets not make such drastic changes so close to the deadline. As such, this should be considered a draft, and not be merged until the Atreus mass production firmware is finalised and tagged.

obra commented:
Ok. Now that we've tagged the Atreus MP firmware, what do we want to do before merging this?

algernon commented:
Take it for a spin, I'd say. I've been using it for about a week now, and it seems to work as expected, but if anyone else could give it a go, see how it feels, that'd be great.

gedankenexperimenter commented:
> The "race" is only a race if both are pressed at the same time (as in, they both get registered in the same scan cycle). That - I think - would be a pretty rare case.

You're thinking of this from the programmer's perspective. To the user, it's a race condition if one perceives the two keypresses as simultaneous, even if the keyscanner doesn't register both presses in the same cycle. If my intent is to press two modifier keys, it doesn't matter in what order the keyboard (or the host) registers them, so I might reasonably think of the two presses as simultaneous. Layer shift keys otherwise behave the same way, from the user's point of view; they change what happens when some other key is pressed, but on their own, they don't (generally) cause anything to happen on the host. But with activation-order combined with transparency, that assumed UI contract is broken.

Without transparency, there would be no problem, but then there would also be no point in having two simultaneously active layer shifts…

gedankenexperimenter commented on src/kaleidoscope/layers.cpp:
```diff
@@ -201,10 +186,16 @@ void Layer_::deactivate(uint8_t layer) {
-  // update top_active_layer_
-  if (layer == top_active_layer_)
-    updateTopActiveLayer();
+  // Rearrange the activation order array...
+  uint8_t idx = 0;
+  for (uint8_t i = active_layer_count_; i > 0; i--) {
+    if (active_layers_[i] == layer) {
+      idx = i;
+      break;
+    }
+  }
+  memmove(&active_layers_[idx], &active_layers_[idx + 1], active_layer_count_ - idx);
+  active_layer_count_--;
```
I may be missing something here, but it looks like there's no protection against deactivating the only layer remaining in the keymap.

algernon replied:
It's a sublte thing, but having now written a test explicitly for this case, I can now say with confidence, that there _is_ protection, just not in `deactivate()`: it's in `updateActiveLayers()`, which starts with `memset(active_layer_keymap_, 0, Runtime.device().numKeys());`. Why is this a protection? It is because `active_layer_keymap_` contains an address->layer mapping for each key on the device. If there are no active layers, we'll default to 0, just like we do prior to this PR.

gedankenexperimenter commented:
I mentioned this in Discord, but I'd also like to go on the record here as having mulled over the user-perceived race condition I mentioned, and come to the conclusion that I don't think it's a reason not to proceed with this change. It's logically equivalent to pressing any two keys at the same time, and for most people I think activation order would make more intuitive sense in that situation, even if the result is (from the user's point of view) non-deterministic.

gedankenexperimenter commented:
Here's a scenario that I'm concerned about (sort of two scenarios):

With index-ordering, it was never possible to truly deactivate the base layer. Even if it was deactivated, keymap lookups would retrieve the base layer's entries if all other layers were off or transparent. With activation-ordering, it could be possible to truly deactivate the base layer, which would constitute a new way for the keymap to get "stuck" in a state that it's impossible to get out of without resetting the keyboard. This could happen even if there are other layers still active.

Secondarily, if all layers get deactivated, Kaleidoscope should have a default layer that activates automatically when `active_layer_count_` goes to zero.

algernon commented:
Okay, this is now rebased on top of master, and I wrote a test case for activation ordering. It also tests the case @gedankenexperimenter highlighted: explicitly turning off the single active layer. We fall back to layer 0 in that case (even if we explicitly turned layer 0 off), just like we do now, without this PR.

eritain commented:
> If my intent is to press two modifier keys, it doesn't matter in what order the keyboard (or the host) registers them, so I might reasonably think of the two presses as simultaneous. Layer shift keys otherwise behave the same way, from the user's point of view; they change what happens when some other key is pressed, but on their own, they don't (generally) cause anything to happen on the host. But with activation-order combined with transparency, that assumed UI contract is broken.

> come to the conclusion that I don't think it's a reason not to proceed

Agreed, but it's a pretty good reason for docs/layers.md to explicitly say, "Note: shifting to a layer is subtly different from using modifier keys." 

To make it act just like multiple modifier keys, where the order you engage them in really is irrelevant, either the user has to precompose a separate layer for every combination (kind of bulky and tedious, the things we want layering to avoid) or the keyboard has to be taught to automatically reorder certain sets of layers whenever they're all together at top of stack (bad idea for multiple reasons). And I think in practice, typists will learn to make sure they engage their layers in deliberate order, the same way we all learned to make sure shift is down before hitting a letter instead of mashing at them near-simultaneously and hoping. So the slight complexity is a reasonable price to pay for the wins layering gives us.

But as to the docs, for a Kaleidoscope noob on a Model 01 (me) it's easy to get the impression that holding fn is essentially like holding shift, ctrl, or alt, except the keyboard keeps *this* bucky bit to itself. It's possible to learn enough to modestly reprogram your fn layer before finding that out; it's just that it throws up a lot of confusion about exactly what fancier behaviors you will or won't be able to program, and how.

The full discussion of "shifting to layers is not like holding modifier keys, order matters" is probably best somewhere under the heading "Layers, transparency, and how lookup works" (when the groundwork is laid to explain why) but it's probably also smart to briefly foreshadow it under "How do I do make layer switching act similar to modifiers?" 

I want to test this PR, but as a Kaleidoscope noob I don't yet have my customizations in version control, so pulling this and merging them into it may involve some fussy copy-paste work first.

algernon commented:
I think that the "Layers, transparency, and how lookup works" section already explains adequatly that order matters:

> It is important to note that transparent keys are looked up from active layers only, from most recently activated to least. [example follows>

And a paragraph later:

> As we just saw, another important factor is that layers are ordered by their order of activation. Whether you activate layer 1 or 2 first, matters.

I can add a short note to the "How do I do make layer switching act similar to modifiers?" section, along these lines:

> While switching layers this way is similar to how modifiers work, there are subtle differences. For a longer explanation, see [later](...)

algernon commented:
Rebased on top of master, and adjusted the documentation a little. The PR also comes with a test case using our googletest-based harness. (The test case, as one would expect, fails if ran with index-ordered layers.)

@obra I'd love if you could take a look! :)

obra commented on tests/features/layers/activation-order/sketch.ino:
```diff
@@ -0,0 +1,98 @@
+  [0] = KEYMAP_STACKED
+  (
+    Key_A ,Key_A ,Key_A ,Key_A ,Key_A ,Key_A ,Key_A
+   ,Key_A ,Key_A ,Key_A ,Key_A ,Key_A ,Key_A ,Key_A
```
Just for clarity, I think it probably makes sense to only have keys we -need- defined on the layer. so just one Key_A.

obra replied:
And -maybe- to have it be Key_0 to represent the layer 0? Not sure on that one.

algernon replied:
`Key_0` makes sense. I went with full-`Key_A` (except for the layer shifts and the macro) to have a uniform layout. But having all the unused keys blank works just as well, I suppose.

gedankenexperimenter replied:
If you want to use a single sketch file for as many testcases as possible (or optimal, given readability considerations), it would be good to have layer-shift keys that are mapped to other keys on the target layer. I think it's reasonable to have some keys transparent, and others (not just one) identify the layer number.

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+}
+
+TEST_F(LayerActivationOrder, LayerZero) {
+  sim_.Press(3, 7); // Macro#0: Layer.deactivate(0)
```
Rather than hard-coding these numbers in the test, here, these should be a preprocessor macro with a descriptive name and the comment should be on the macro.

gedankenexperimenter replied:
How about a constexpr instead, to encourage better habits elsewhere?

algernon replied:
Yup, I'm going with constexpr. I'm going to touch the harness too, because `sim_.Press()` & friends currently take a pair of `uint8_t`, need to teach them to have a `KeyAddr` variant too.

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+  sim_.Press(3, 7); // Macro#0: Layer.deactivate(0)
+  auto state = RunCycle();
+
+  sim_.Press(0, 0); // A
```
ditto - prerocessor macro

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+  // states, we shouldn't emit a report.
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 0);
+
+  // Pressing (0, 1), the report shall contain 'A' _and_ 'B'.
```
I think this may want some more documentation about -why-. It's hinted at in the title of the test, but their CamelCase requirement makes it non-obvious.

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_A));
```
We have this pattern repeated so many times. I feel like it should get extracted to a function (even if it's only in this test file)

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+
+class LayerActivationOrder : public VirtualDeviceTest {
+ public:
+  void SingleKeyTest(Key k) {
```
Can this get a more descriptive name?  maybe something active. 
And I know that all our tests have the key at 0,0, but maybe we want to pass it in as a param anyway.

TestPressAndRelease(KEYSWITCH_TOP_LEFT, Key_A); ?

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,177 @@
+  SingleKeyTest(Key_B);
+
+  // Releasing (3,6) gets us back to the base layer
+  sim_.Release(3, 6); // ShiftToLayer(1)
```
preprocessor macro

obra reviewed (changes requested):
I've got a bunch of very nit-picky comments about the form of the tests, mostly because these are going to end up getting cargo-culted.

algernon commented:
Pushed an update, the testcase is almost completely rewritten. I lifted out the common parts, gave them new names, and am using constants throughout the test, instead of hardcoding positions and keys all over the place.

algernon commented:
Hm. I need to update the removal dates. They're in the past at the moment.

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,206 @@
+    sim_.Release(addr.row(), addr.col());
+  }
+
+  auto pressKeyswitchAndCycle(const KeyAddr& addr) {
```
AndRunCycle?

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,206 @@
+    return RunCycle();
+  }
+
+  void assertSingleReportThatContains(std::unique_ptr<State> &state, Key k) {
```
assertSingleReportContaining

obra commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,206 @@
+      Contains(k));
+  }
+
+  void assertSingleReportThatDoesNotContain(std::unique_ptr<State> &state, Key k) {
```
assertSingleReportNotContaining

obra commented:
This feels a lot nicer. A couple naming things popped out, but that was it.

algernon commented:
Updated, now it has the new names :)

obra commented:
Yay! Thank you so much, @algernon

gedankenexperimenter commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,206 @@
+
+class LayerActivationOrder : public VirtualDeviceTest {
+ public:
+  const KeyAddr KEYSWITCH_TOP_LEFT = KeyAddr{0, 0};   // layer-dependent key
+  const KeyAddr KEYSWITCH_TOP_RIGHT = KeyAddr{0, 15}; // layer-dependent key
+  const KeyAddr KEYSWITCH_LEFT_PALM = KeyAddr{3, 6};  // ShiftToLayer(1)
+  const KeyAddr KEYSWITCH_RIGHT_PALM = KeyAddr{3, 9}; // ShiftToLayer(2)
+  const KeyAddr KEYSWITCH_LEFT_THUMB_RIGHTMOST = KeyAddr{3, 7}; // L0 deactivate macro
+
+  const Key LAYER0_KEY = Key_0;
+  const Key LAYER1_KEY = Key_1;
+  const Key LAYER2_KEY = Key_2;
```
[begin nitpick]

The [coding style guide](https://github.com/keyboardio/Kaleidoscope/blob/master/docs/codebase/code-style.md#constant-names) would have us naming these constants things like `kKeyswitchTopLeft` instead, and reserving all-caps for preprocessor macros. I'm ambivalent about this, and actually tend to just name contants like other variables, but I would like to get a ruling on the preferred style from those in authority, and perhaps an update to the style guide.

On a separate note, I think it would be preferable for the first group to be in namespace `kaleidoscope::testing`, and to be defined like this:
```
constexpr KeyAddr KEYSWITCH_TOP_LEFT{0, 0};
```
…which would be a more C++ style brace initialization.

[end nitpick]

algernon replied:
The coding style guide is... still somewhat a work in progress - I need to open specific issues so we can track what needs to be updated still. We don't prefix constants with `k` anywhere in existing code, for example (outside of googletest).

> constexpr KeyAddr KEYSWITCH_TOP_LEFT{0, 0};

We'd need to make it static then, and then have a `constexpr KeyAddr LayerActivationOrder::KEYSWITCH_TOP_LEFT;` outside of the class too, and the same for every other constant. While it would be more C++, and more in line with how we usually do things in Kaleidoscope, it doesn't really make the code any more readable, imo.

gedankenexperimenter replied:
I'm well aware that the style guide isn't followed everywhere (or, in some cases, anywhere). I'm arguing that it would be real nice for someone who wants to follow a consistent style to have a decision from the people who are in authority, so I can stop having a mental debate every time I have to name a constant. And using all caps for constants in a file with lots of preprocessor macros is distracting.

Re: namespace: I'm saying that these constants aren't just for class `LayerActivationOrder`, so they should be defined outside the class. Preferably in a place (in the case of that first block) where they could be included by other tests, as well. It only takes one line to define each constant, in the `testing` namespace instead of in the class; nothing else would change.

gedankenexperimenter commented on tests/features/layers/activation-order/test/testcase.cpp:
```diff
@@ -0,0 +1,206 @@
+  const Key LAYER1_KEY = Key_1;
+  const Key LAYER2_KEY = Key_2;
+
+  void pressKeyswitch(const KeyAddr& addr) {
```
Why is the parameter a reference here, rather than just a `KeyAddr`?
